File: myExpenses/src/test/java/org/totschnig/myexpenses/util/UiUtilsDateModeTest.java
package org.totschnig.myexpenses.util;

import junit.framework.Assert;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mockito;
import org.robolectric.RobolectricTestRunner;
import org.totschnig.myexpenses.model.Account;
import org.totschnig.myexpenses.model.AccountType;
import org.totschnig.myexpenses.preference.PrefHandler;
import org.totschnig.myexpenses.preference.PrefKey;

import static org.mockito.ArgumentMatchers.anyBoolean;
import static org.mockito.ArgumentMatchers.eq;
import static org.totschnig.myexpenses.preference.PrefKey.TRANSACTION_WITH_TIME;
import static org.totschnig.myexpenses.preference.PrefKey.TRANSACTION_WITH_VALUE_DATE;
import static org.totschnig.myexpenses.util.UiUtils.DateMode.BOOKING_VALUE;
import static org.totschnig.myexpenses.util.UiUtils.DateMode.DATE;
import static org.totschnig.myexpenses.util.UiUtils.DateMode.DATE_TIME;

@RunWith(RobolectricTestRunner.class)
public class UiUtilsDateModeTest {
  private Account cashAccount, bankAcount;
  private PrefHandler prefHandler;

  @Before
  public void setup() {
    cashAccount = Mockito.mock(Account.class);
    bankAcount = Mockito.mock(Account.class);
    Mockito.when(cashAccount.getType()).thenReturn(AccountType.CASH);
    Mockito.when(bankAcount.getType()).thenReturn(AccountType.BANK);
    prefHandler = Mockito.mock(PrefHandler.class);
  }

  private void mockPref(PrefKey prefKey, boolean value) {
    Mockito.when(prefHandler.getBoolean(eq(prefKey), anyBoolean())).thenReturn(value);
  }

  @Test
  public void withTimeAndWithValueDate() {
    mockPref(TRANSACTION_WITH_TIME, true);
    mockPref(TRANSACTION_WITH_VALUE_DATE, true);

    Assert.assertEquals(DATE_TIME, dateMode(cashAccount));
    Assert.assertEquals(BOOKING_VALUE, dateMode(bankAcount));
  }

  @Test
  public void withTimeAndWithoutValueDate() {
    mockPref(TRANSACTION_WITH_TIME, true);
    mockPref(TRANSACTION_WITH_VALUE_DATE, false);

    Assert.assertEquals(DATE_TIME, dateMode(cashAccount));
    Assert.assertEquals(DATE_TIME, dateMode(bankAcount));
  }

  @Test
  public void withoutTimeAndWithValueDate() {
    mockPref(TRANSACTION_WITH_TIME, false);
    mockPref(TRANSACTION_WITH_VALUE_DATE, true);

    Assert.assertEquals(DATE, dateMode(cashAccount));
    Assert.assertEquals(BOOKING_VALUE, dateMode(bankAcount));
  }

  @Test
  public void withoutTimeAndWithoutValueDate() {
    mockPref(TRANSACTION_WITH_TIME, false);
    mockPref(TRANSACTION_WITH_VALUE_DATE, false);

    Assert.assertEquals(DATE, dateMode(cashAccount));
    Assert.assertEquals(DATE, dateMode(bankAcount));
  }

  private UiUtils.DateMode dateMode(Account account) {
    return  UiUtils.getDateMode(account.getType(), prefHandler);
  }
}

File: myExpenses/src/test/java/org/totschnig/myexpenses/util/UiUtilsDateModeTest.kt
package org.totschnig.myexpenses.util

import com.google.common.truth.Truth.assertThat
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.kotlin.any
import org.mockito.kotlin.eq
import org.mockito.kotlin.whenever
import org.robolectric.RobolectricTestRunner
import org.totschnig.myexpenses.model.AccountType
import org.totschnig.myexpenses.prefHandler
import org.totschnig.myexpenses.preference.PrefKey
import org.totschnig.myexpenses.util.ui.UiUtils.DateMode
import org.totschnig.myexpenses.util.ui.getDateMode

@RunWith(RobolectricTestRunner::class)
class UiUtilsDateModeTest {

    private fun mockPref(prefKey: PrefKey, value: Boolean) {
       whenever(prefHandler.getBoolean(eq(prefKey), any())).thenReturn(value)
    }

    @Test
    fun withTimeAndWithValueDate() {
        mockPref(PrefKey.TRANSACTION_WITH_TIME, true)
        mockPref(PrefKey.TRANSACTION_WITH_VALUE_DATE, true)
        assertThat(dateMode(AccountType.CASH)).isEqualTo(DateMode.DATE_TIME)
        assertThat(dateMode(AccountType.BANK)).isEqualTo(DateMode.BOOKING_VALUE)
    }

    @Test
    fun withTimeAndWithoutValueDate() {
        mockPref(PrefKey.TRANSACTION_WITH_TIME, true)
        mockPref(PrefKey.TRANSACTION_WITH_VALUE_DATE, false)
        assertThat(dateMode(AccountType.CASH)).isEqualTo(DateMode.DATE_TIME)
        assertThat(dateMode(AccountType.BANK)).isEqualTo(DateMode.DATE_TIME)
    }

    @Test
    fun withoutTimeAndWithValueDate() {
        mockPref(PrefKey.TRANSACTION_WITH_TIME, false)
        mockPref(PrefKey.TRANSACTION_WITH_VALUE_DATE, true)
        assertThat(dateMode(AccountType.CASH)).isEqualTo(DateMode.DATE)
        assertThat(dateMode(AccountType.BANK)).isEqualTo(DateMode.BOOKING_VALUE)
    }

    @Test
    fun withoutTimeAndWithoutValueDate() {
        mockPref(PrefKey.TRANSACTION_WITH_TIME, false)
        mockPref(PrefKey.TRANSACTION_WITH_VALUE_DATE, false)
        assertThat(dateMode(AccountType.CASH)).isEqualTo(DateMode.DATE)
        assertThat(dateMode(AccountType.BANK)).isEqualTo(DateMode.DATE)
    }

    private fun dateMode(accountType: AccountType): DateMode {
        return getDateMode(accountType, prefHandler)
    }
}